{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RiboSeq processing pipeline \n",
    "\n",
    "#### Steps\n",
    "\n",
    "This pipe assumes you have the fastq files to process in a dedicated directory within a subdirectory called data.\n",
    "\n",
    "1) Run Fastqc to check the read length distribution.  Size selection was used during the experiment, expect both 26-34nt for monosome and 54-68nt for disome according to Erica's email.\n",
    "\n",
    "2) Run cutadapt using the parameters provided by Ezrabio.    \"-j 8 -g \"^GGG\" -a \"A{10}\" -n 2 -m 15 --max-n=0.1 --discard-casava -o output.fastq.gz input.fastq.gz\"\n",
    "\n",
    "3) Remove reads where the first position quality score is <=10\n",
    "\n",
    "4) Align reads with Bowtie to non-coding RNA, https://downloads.yeastgenome.org/sequence/S288C_reference/rna/archive/rna_coding_R64-1-1_20110203.fasta.gz reads that align will be discarded.  Allow 1 mismatch in bowtie alignment.\n",
    "\n",
    "Alignment files (SAM/BAM) provide forward/reverse alignment information is provided by the Sam Flag.  \n",
    "    0 is read aligned in the Fwd direction\n",
    "    4 is unaligned\n",
    "    16 is aligned in the Rvs direction\n",
    "\n",
    "5) Align the remaining reads with Bowtie to YPS1009 and S288C reference genomes.\n",
    "\n",
    "6) Run samtools mpileup to generate base counts for all genes in YPS1009 and S288C.\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Samples labeled as -RNA are strictly RNA-Seq and will be processed with the normal condor RNA-Seq pipeline.\n",
    "\n",
    "/mnt/bigdata/linuxhome/mplace/data/Ribo-seq/Leah-12142023/RNA-Seq\n",
    "\n",
    "~/scripts/Condor-RNA-Seq-Pipeline/rnaSeqCondor_paired.py -e mplace@wisc.edu -f filelist -ref R64\n",
    "\n",
    "~/scripts/Condor-RNA-Seq-Pipeline/rnaSeqCondor_paired.py -e mplace@wisc.edu -f filelist -ref YPS1009"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/bigdata/linuxhome/mplace/data/Ribo-seq/Leah-12142023/\n"
     ]
    }
   ],
   "source": [
    "# import required Python modules\n",
    "import glob\n",
    "import itertools\n",
    "import os\n",
    "import re\n",
    "import subprocess\n",
    "import sys\n",
    "\n",
    "# set global variables\n",
    "parentDir = os.getcwd() + \"/\"\n",
    "referenceDir = \"/mnt/bigdata/linuxhome/mplace/scripts/riboSeqPipeline/reference/\"\n",
    "print(parentDir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create the list of fastq files to process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a list file of fastq files for processing, we are assuming the original fastq files are in a directory called data.\n",
    "dataDir =  parentDir + \"data/\"        \n",
    "with open('inputFastq.txt', 'w') as out:\n",
    "    for fstq in glob.glob(dataDir + \"*.fastq\"):\n",
    "        out.write(fstq + \"\\n\")               "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Write fastqc condor submit files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 1)\n",
    "# write the fastqc condor submit file\n",
    "with open('fastqc.submit', 'w') as submit:\n",
    "            submit.write( \"Universe                 = vanilla\\n\" )\n",
    "            submit.write( \"Executable               = runFastqc.sh\\n\")\n",
    "            submit.write( \"Arguments                = $(fastqFile)\\n\")\n",
    "            submit.write( \"Error                    = fastqc.submit.err\\n\")\n",
    "            submit.write( \"Log                      = fastqc.submit.log\\n\")  \n",
    "            submit.write( \"Requirements             = OpSysandVer == \\\"CentOS7\\\"\\n\")\n",
    "            submit.write( \"Queue fastqFile from inputFastq.txt\\n\" )\n",
    "submit.close()  \n",
    "\n",
    "# write shell script to run fastqc\n",
    "with open('runFastqc.sh', 'w') as out:\n",
    "    out.write(\"#!/bin/bash\\n\")\n",
    "    out.write(\"source /opt/bifxapps/miniconda3/etc/profile.d/conda.sh\\n\")\n",
    "    out.write(\"unset PYTHONPATH\\n\")  \n",
    "    out.write(\"conda activate /home/glbrc.org/mplace/.conda/envs/riboSeq\\n\")\n",
    "    out.write(\"fastqc $1\\n\")\n",
    "    out.write(\"conda deactivate\")\n",
    "out.close()\n",
    "\n",
    "os.chmod('runFastqc.sh', 0o0777)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Using the GLBRC condor submit node submit the above fastq.submit file\n",
    "\n",
    " condor_submit fastqc.submit  \n",
    "\n",
    " check that the job is running\n",
    "\n",
    " condor_q , you should see something like the following\n",
    "\n",
    "Schedd: scarcity-submit.glbrc.org : <144.92.98.21:9618?... @ 12/04/23 14:27:03\n",
    "\n",
    "OWNER  BATCH_NAME    SUBMITTED   DONE   RUN    IDLE  TOTAL JOB_IDS\n",
    "\n",
    "mplace ID: 814372  12/4  14:25      _      _      2      2 814372.0-1\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Prepare to run Cutadapt "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 2)\n",
    "# setup input file for cutadapt\n",
    "cutadaptOutDir = parentDir + 'cutadapt/'\n",
    "if os.path.exists(cutadaptOutDir):\n",
    "    print(\"Directory exists.\")\n",
    "else:\n",
    "    os.mkdir(cutadaptOutDir)\n",
    "\n",
    "with open('inputFastq.txt', 'r') as f, open('cutadaptInput.txt', 'w') as out:\n",
    "    for fstq in f:\n",
    "        fstqName = re.sub('.fastq', '-clean.fastq', os.path.basename(fstq.rstrip()))\n",
    "        fstqOutName = cutadaptOutDir + fstqName\n",
    "        out.write(f'{fstq.rstrip()} {fstqOutName}\\n')\n",
    "f.close()\n",
    "out.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Write cutadapt condor submit files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write the cutadapt condor submit file\n",
    "with open('cutadapt.submit', 'w') as submit:\n",
    "    submit.write( \"Universe                 = vanilla\\n\" )\n",
    "    submit.write( \"Executable               = runCutAdapt.sh\\n\")\n",
    "    submit.write( \"Arguments                = $(fastqFile) $(outFastq)\\n\")\n",
    "    submit.write( \"Error                    = cutadapt.submit.err\\n\")\n",
    "    submit.write( \"Log                      = cutadapt.submit.log\\n\")  \n",
    "    submit.write( \"Requirements             = OpSysandVer == \\\"CentOS7\\\"\\n\")\n",
    "    submit.write( \"Queue fastqFile, outFastq from cutadaptInput.txt\\n\" )\n",
    "submit.close()\n",
    "\n",
    "# write shell script to run cutadapt\n",
    "with open('runCutAdapt.sh', 'w') as out:\n",
    "    out.write(\"#!/bin/bash\\n\")\n",
    "    out.write(\"source /opt/bifxapps/miniconda3/etc/profile.d/conda.sh\\n\")\n",
    "    out.write(\"unset PYTHONPATH\\n\")  \n",
    "    out.write(\"conda activate /home/glbrc.org/mplace/.conda/envs/riboSeq\\n\")\n",
    "    out.write(\"cutadapt -j 8 -g ^GGG -a A{10} -n 2 -m 15 --max-n=0.1 --discard-casava -o $2 $1\\n\")\n",
    "    out.write(\"conda deactivate\")\n",
    "out.close()\n",
    "\n",
    "os.chmod('runCutAdapt.sh', 0o0777)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  on scarcity-submit run the above cutadapt condor files\n",
    "\n",
    "condor_submit cutadapt.submit"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Next remove reads where the 1st position quality score is low.\n",
    "### This only removes a handful of reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 3) Remove reads where the first position quality score is <=10\n",
    "\n",
    "qualScores = ['+', '*', ')','(', \"'\", '&', '%', '$', '#', '\"', '!']\n",
    "for fastq in glob.glob('cutadapt/*-clean.fastq'):\n",
    "    name = re.sub('-clean.fastq', '-filt.fastq', fastq)\n",
    "    with open(fastq, 'r') as f, open(name, 'w') as out:\n",
    "        for hdr, seq, plus, qual in itertools.zip_longest(*[f]*4):\n",
    "            firstQual = [*qual][0]\n",
    "            if not firstQual in qualScores:\n",
    "                out.write(hdr)\n",
    "                out.write(seq)\n",
    "                out.write(plus)\n",
    "                out.write(qual)\n",
    "            else:\n",
    "                print(hdr, '  ', qual)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Align reads to Non-coding RNA using Bowtie2.  \n",
    "### Identifies which reads to remove from further consideration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 4 ) Align reads with Bowtie2 to non-coding RNA, \n",
    "# https://downloads.yeastgenome.org/sequence/S288C_reference/rna/archive/rna_coding_R64-1-1_20110203.fasta.gz \n",
    "# reads that align will be discarded.  Allow 1 mismatch in bowtie2 alignment.\n",
    "# bowtie2 -p 8 --phred33 -N 1 -x $REFERENCE -U $file -S $out.sam \n",
    "# -p number of threads\n",
    "# -N Sets the number of mismatches\n",
    "# -x The basename of the index for the reference genome\n",
    "# -U file to align (unpaired)\n",
    "# -S File to write SAM alignments to\n",
    "\n",
    "# setup input file for bowtie2 alignment to non-coding RNA\n",
    "nonCodingOutDir = parentDir + 'alignNonCodingRNA/'\n",
    "if os.path.exists(nonCodingOutDir):\n",
    "    print(\"Directory exists.\")\n",
    "else:\n",
    "    os.mkdir(nonCodingOutDir)\n",
    "\n",
    "# get a list of cutadapt cleaned & filtered fastq files for alignment\n",
    "with open('alignmentInput.txt', 'w') as out:\n",
    "    for cleanfstq in glob.glob(cutadaptOutDir + '*-filt.fastq'):\n",
    "        samFile = re.sub('cutadapt', 'alignNonCodingRNA', re.sub('-filt.fastq', '.sam', cleanfstq))        \n",
    "        out.write(cleanfstq + ' ' + samFile + '\\n')\n",
    "out.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Write Bowtie2 condor submit files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write the bowtie2 condor submit file\n",
    "with open('ncbowtie2.submit', 'w') as submit:\n",
    "    submit.write( \"Universe                 = vanilla\\n\" )\n",
    "    submit.write( \"Executable               = runncBowtie2.sh\\n\")\n",
    "    submit.write( \"Arguments                = $(fastqFile) $(sam)\\n\")\n",
    "    submit.write( \"Error                    = ncbowtie2.submit.err\\n\")\n",
    "    submit.write( \"Log                      = ncbowtie2.submit.log\\n\")  \n",
    "    submit.write( \"Requirements             = OpSysandVer == \\\"CentOS7\\\"\\n\")\n",
    "    submit.write( \"Queue fastqFile, sam from alignmentInput.txt\\n\" )\n",
    "submit.close()\n",
    "\n",
    "# write shell script to run cutadapt\n",
    "with open('runncBowtie2.sh', 'w') as out:\n",
    "    out.write(\"#!/bin/bash\\n\")\n",
    "    out.write(\"source /opt/bifxapps/miniconda3/etc/profile.d/conda.sh\\n\")\n",
    "    out.write(\"unset PYTHONPATH\\n\")  \n",
    "    out.write(\"conda activate /home/glbrc.org/mplace/.conda/envs/riboSeq\\n\")\n",
    "    out.write(\"bowtie2 -p 8 --phred33 -N 1 -x /mnt/bigdata/linuxhome/mplace/scripts/riboseq/reference/rna_coding_R64-1-1 -U $1 -S $2\\n\")\n",
    "    out.write(\"conda deactivate\")\n",
    "out.close()\n",
    "\n",
    "os.chmod('runncBowtie2.sh', 0o0777)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  on scarcity-submit run the above condor files\n",
    "\n",
    "condor_submit ncbowtie2.submit"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove reads that aligned to Non-coding using samtools view.\n",
    "#### -F --exclude-flags FLAG ,     Do not output alignments with any bits set in FLAG present in the FLAG field.\n",
    "#### in this case we exclude the reads which did not align to the Non-coding RNA i.e. the reads we want to analyze later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get a list of reads which aligned to Non-Coding RNA and remove them from the clean.fastq files\n",
    "#samtools view -F 4 -u  -O SAM -o mapped.sam TestSample1.sam\n",
    "with open('filterSamInput.txt', 'w') as out:\n",
    "    for sam in glob.glob(nonCodingOutDir + '*.sam'):\n",
    "        sampleName = re.sub('.sam', '', os.path.basename(sam))\n",
    "        outSam = nonCodingOutDir + sampleName + '-unmapped.sam'\n",
    "        out.write(sam + ' ' + outSam + '\\n')\n",
    "out.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Write the samtools condor submit file to create the clean fastq files for the next step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write the samtools filter UNMAPPED (reads which did not align to Non-Coding RNA) reads condor submit file\n",
    "# these reads will be aligned to S288C and YPS1009\n",
    "with open('filter.submit', 'w') as submit:\n",
    "    submit.write( \"Universe                 = vanilla\\n\" )\n",
    "    submit.write( \"Executable               = runfilter.sh\\n\")\n",
    "    submit.write( \"Arguments                = $(sam) $(ncsam)\\n\")\n",
    "    submit.write( \"Error                    = filter.submit.err\\n\")\n",
    "    submit.write( \"Log                      = filter.submit.log\\n\")  \n",
    "    submit.write( \"Requirements             = OpSysandVer == \\\"CentOS7\\\"\\n\")\n",
    "    submit.write( \"Queue sam, ncsam from filterSamInput.txt\\n\" )\n",
    "submit.close()\n",
    "\n",
    "# write shell script to run cutadapt\n",
    "with open('runfilter.sh', 'w') as out:\n",
    "    out.write(\"#!/bin/bash\\n\")\n",
    "    out.write(\"source /opt/bifxapps/miniconda3/etc/profile.d/conda.sh\\n\")\n",
    "    out.write(\"unset PYTHONPATH\\n\")  \n",
    "    out.write(\"conda activate /home/glbrc.org/mplace/.conda/envs/riboSeq\\n\")\n",
    "    out.write(\"samtools view -f 4 -u -O SAM -o $2 $1\\n\")\n",
    "    out.write(\"conda deactivate\")\n",
    "out.close()\n",
    "\n",
    "os.chmod('runfilter.sh', 0o0777)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  on scarcity-submit run the above condor files\n",
    "\n",
    "condor_submit filter.submit"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create the alignment results directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create output files for genome alignments\n",
    "if not os.path.exists(parentDir + 'alignments'):\n",
    "    os.mkdir(parentDir + 'alignments')\n",
    "    os.mkdir(parentDir + 'alignments/S288C')\n",
    "    os.mkdir(parentDir + 'alignments/YPS1009')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create new fastq files with the reads which did not map to the Non-Coding RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create new fastq files by filtering for reads that DID NOT ALIGN to the Non-Coding RNA\n",
    "# First create a new file containing the read names (for reads we want to keep)\n",
    "# nonCodingOutDir\n",
    "for unmapped in glob.glob(nonCodingOutDir + '*-unmapped.sam'):\n",
    "    nameFile = re.sub('-unmapped.sam', '', os.path.basename(unmapped))\n",
    "    outFile  = parentDir + 'alignments/' + nameFile + '-names.txt'\n",
    "    with open(unmapped) as f, open(outFile, 'w') as out:\n",
    "        for line in f:\n",
    "            name = line.split('\\t')[0]\n",
    "            out.write(f'{name}\\n')\n",
    "    f.close()\n",
    "    out.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Use seqtk to subset the unmapped reads for use with bowtie2\n",
    "for fstq in glob.glob(cutadaptOutDir + '*-filt.fastq'):\n",
    "    print('processing: ', fstq)\n",
    "    outFile = parentDir + 'alignments/' +  re.sub('-filt.fastq', '.fastq', os.path.basename(fstq))    # create output file name\n",
    "    nameLst = parentDir + 'alignments/' + re.sub('-filt.fastq', '-names.txt', os.path.basename(fstq)) \n",
    "    cmd = [ 'seqtk', 'subseq', fstq, nameLst ]\n",
    "    # run command and capture output\n",
    "    output = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE).communicate()    \n",
    "    # write results to file\n",
    "    with open(outFile, 'w') as out:\n",
    "        out.write(output[0].decode('utf-8'))\n",
    "    out.close()     "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set up for alignments to S288C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create input file for alignment to the S288C reference genome.\n",
    "outputPath = parentDir + 'alignments/S288C/'\n",
    "with open('refAlignmentS288C_input.txt', 'w') as out:\n",
    "    for inFastq in glob.glob(parentDir + 'alignments/*.fastq'):\n",
    "        sampleName = re.sub('.fastq', '.sam', os.path.basename(inFastq))\n",
    "        out.write(inFastq + ' ' + outputPath + sampleName + ' /home/glbrc.org/mplace/data/reference/S288C_reference_genome_R64-1-1_20110203/s.cerevisiae-R64-1-1' + '\\n')\n",
    "out.close"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create input file for alignment to the YPS1009 reference genome.\n",
    "outputPath = parentDir + 'alignments/YPS1009/'\n",
    "with open('refAlignmentYPS1009_input.txt', 'w') as out:\n",
    "    for inFastq in glob.glob(parentDir + 'alignments/*.fastq'):\n",
    "        sampleName = re.sub('.fastq', '.sam', os.path.basename(inFastq))\n",
    "        out.write(inFastq + ' ' + outputPath + sampleName + ' /home/glbrc.org/mplace/data/reference/YPS1009/YPS1009' + ' \\n')\n",
    "out.close"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Write the condor submit files for running bowtie2 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 5) Align reads to reference genomes S288C and YPS1009\n",
    "# align reads to S288C using bowtie2 \n",
    "# write the bowtie2 condor submit file\n",
    "#S288C\n",
    "with open('s288cbowtie2.submit', 'w') as submit:\n",
    "    submit.write( \"Universe                 = vanilla\\n\" )\n",
    "    submit.write( \"Executable               = runBowtie2.sh\\n\")\n",
    "    submit.write( \"Arguments                = $(fastqFile) $(sam) $(ref)\\n\")\n",
    "    submit.write( \"Error                    = s288c_bowtie2.submit.err\\n\")\n",
    "    submit.write( \"Log                      = s288c_bowtie2.submit.log\\n\")  \n",
    "    submit.write( \"Requirements             = OpSysandVer == \\\"CentOS7\\\"\\n\")\n",
    "    submit.write( \"Queue fastqFile, sam, ref from refAlignmentS288C_input.txt\\n\" )\n",
    "submit.close()\n",
    "#YPS1009\n",
    "with open('yps1009bowtie2.submit', 'w') as submit:\n",
    "    submit.write( \"Universe                 = vanilla\\n\" )\n",
    "    submit.write( \"Executable               = runBowtie2.sh\\n\")\n",
    "    submit.write( \"Arguments                = $(fastqFile) $(sam) $(ref)\\n\")\n",
    "    submit.write( \"Error                    = yps1009_bowtie2.submit.err\\n\")\n",
    "    submit.write( \"Log                      = yps1009_bowtie2.submit.log\\n\")  \n",
    "    submit.write( \"Requirements             = OpSysandVer == \\\"CentOS7\\\"\\n\")\n",
    "    submit.write( \"Queue fastqFile, sam, ref from refAlignmentYPS1009_input.txt\\n\" )\n",
    "submit.close()\n",
    "\n",
    "# write shell script to run bowtie2\n",
    "with open('runBowtie2.sh', 'w') as out:\n",
    "    out.write(\"#!/bin/bash\\n\")\n",
    "    out.write(\"source /opt/bifxapps/miniconda3/etc/profile.d/conda.sh\\n\")\n",
    "    out.write(\"unset PYTHONPATH\\n\")  \n",
    "    out.write(\"conda activate /home/glbrc.org/mplace/.conda/envs/riboSeq\\n\")\n",
    "    out.write(f\"bowtie2 -p 8 --phred33 -N 1 -x $3 -U $1 -S $2\\n\")\n",
    "    out.write(\"conda deactivate\")\n",
    "out.close()\n",
    "\n",
    "os.chmod('runBowtie2.sh', 0o0777)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  on scarcity-submit run the above condor files\n",
    "\n",
    "condor_submit s288cbowtie2.submit\n",
    "\n",
    "condor_submit yps1009bowtie2.submit"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sort the alignment files for the next step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# samtools sort ANEU-LOG-30_S2_R1_001.sam > ANEU-LOG-30_S2_R1_001-sorted.sam\n",
    "# samtools mpileup -f ~/data/reference/S288C_reference_genome_R64-1-1_20110203/S288C_reference_sequence_R64-1-1_20110203.fsa ANEU-LOG-30_S2_R1_001-sorted.sam > ANEU-LOG-30_pileup.txt\n",
    "# Create input file for samtools sort \n",
    "inputFilePath = parentDir + 'alignments/YPS1009/'\n",
    "print(inputFilePath)\n",
    "with open('sort-YPS1009_input.txt', 'w') as out:\n",
    "    for sam in glob.glob(inputFilePath + '*.sam'):\n",
    "        sampleName = re.sub('.sam', '.bam', os.path.basename(sam))\n",
    "        out.write(sam + ' ' + inputFilePath + sampleName + ' \\n')\n",
    "out.close"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "inputFilePath = parentDir + 'alignments/S288C/'\n",
    "print(inputFilePath)\n",
    "with open('sort-S288C_input.txt', 'w') as out:\n",
    "    for sam in glob.glob(inputFilePath + '*.sam'):\n",
    "        sampleName = re.sub('.sam', '.bam', os.path.basename(sam))\n",
    "        out.write(sam + ' ' + inputFilePath + sampleName + ' \\n')\n",
    "out.close"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Write the condor submit files for sorting the alignment files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sort the alignment output files S288C and YPS1009\n",
    "with open('sort.submit', 'w') as submit:\n",
    "    submit.write( \"Universe                 = vanilla\\n\" )\n",
    "    submit.write( \"Executable               = runsort.sh\\n\")\n",
    "    submit.write( \"Arguments                = $(sam) $(bam)\\n\")\n",
    "    submit.write( \"Error                    = sort.submit.err\\n\")\n",
    "    submit.write( \"Log                      = sort.submit.log\\n\")  \n",
    "    submit.write( \"Requirements             = OpSysandVer == \\\"CentOS7\\\"\\n\")\n",
    "    #submit.write( \"Queue sam, bam from  sort-S288C_input.txt\\n\" )\n",
    "    submit.write( \"Queue sam, bam from sort-YPS1009_input.txt\\n\" )\n",
    "submit.close()\n",
    "\n",
    "# write shell script to run sorting\n",
    "with open('runsort.sh', 'w') as out:\n",
    "    out.write(\"#!/bin/bash\\n\")\n",
    "    out.write(\"source /opt/bifxapps/miniconda3/etc/profile.d/conda.sh\\n\")\n",
    "    out.write(\"unset PYTHONPATH\\n\")  \n",
    "    out.write(\"conda activate /home/glbrc.org/mplace/.conda/envs/riboSeq\\n\")\n",
    "    out.write(\"samtools sort $1  > $2\\n\")\n",
    "    out.write(\"conda deactivate\")\n",
    "out.close()\n",
    "\n",
    "os.chmod('runsort.sh', 0o0777)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create a dictionary of S288C chromosome sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ref|NC_001133|': 230218, 'ref|NC_001134|': 813184, 'ref|NC_001135|': 316620, 'ref|NC_001136|': 1531933, 'ref|NC_001137|': 576874, 'ref|NC_001138|': 270161, 'ref|NC_001139|': 1090940, 'ref|NC_001140|': 562643, 'ref|NC_001141|': 439888, 'ref|NC_001142|': 745751, 'ref|NC_001143|': 666816, 'ref|NC_001144|': 1078177, 'ref|NC_001145|': 924431, 'ref|NC_001146|': 784333, 'ref|NC_001147|': 1091291, 'ref|NC_001148|': 948066, 'ref|NC_001224|': 85779}\n"
     ]
    }
   ],
   "source": [
    "# S288C chromosome sizes\n",
    "S288C_chromSizes = {}\n",
    "with open('/mnt/bigdata/linuxhome/mplace/data/reference/S288C_reference_genome_R64-1-1_20110203/individualChroms/S288C.chromsizes', 'r') as f:\n",
    "    for ln in f:\n",
    "        chrom, length = ln.rstrip().split()\n",
    "        if chrom not in S288C_chromSizes and chrom != '2-micron':\n",
    "            S288C_chromSizes[chrom] = int(length)\n",
    "\n",
    "print(S288C_chromSizes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create information dictionaries for S288C\n",
    "gff = '/mnt/bigdata/linuxhome/mplace/data/reference/S288C_reference_genome_R64-1-1_20110203/saccharomyces_cerevisiae_R64-1-1_20110208_noFasta.gff'\n",
    "geneLookUp = {}    # dictionary of dictionaries key = chrom : {key = the -72 position of gene value = gene name }\n",
    "gffDict = {}       # dictionary of dictionaries key = chrom : {key = {start : 0, end : 0, strand : '+'}\n",
    "\n",
    "with open(gff, 'r') as g:\n",
    "    for line in g:\n",
    "        if line.startswith('#'):   # skip comment rows\n",
    "            continue\n",
    "        else:\n",
    "            dat = line.split('\\t')\n",
    "            chrom = dat[0]\n",
    "            if chrom not in S288C_chromSizes:\n",
    "                continue\n",
    "            if chrom not in geneLookUp:\n",
    "                geneLookUp[chrom] = {}\n",
    "            # add chrom to gffDict\n",
    "            if chrom not in gffDict:\n",
    "                gffDict[chrom] = {}\n",
    "            # identify genes\n",
    "            if dat[2] == 'gene':\n",
    "                geneName = re.sub('ID=', '', dat[8].split(';')[0])\n",
    "                # identify strand\n",
    "                if dat[6] == '+':          # POSIIVE STRAND\n",
    "                    if int(dat[3]) < 72:          \n",
    "                        start = 0\n",
    "                    else:\n",
    "                        start = int(dat[3]) - 72\n",
    "                    if start not in geneLookUp:\n",
    "                        geneLookUp[chrom][start] = geneName\n",
    "                    else:\n",
    "                        print('Duplicate start ', geneName, geneLookUp[chrom][start])\n",
    "\n",
    "                    if not int(dat[4]) + 60 > S288C_chromSizes[chrom]:\n",
    "                        end   = int(dat[4]) + 60\n",
    "                    else:\n",
    "                        end = S288C_chromSizes[chrom]\n",
    "\n",
    "                    # add final gene info for positive strand \n",
    "                    if geneName not in gffDict[chrom]:\n",
    "                        gffDict[chrom][geneName] = {'strand' : '+', 'start': start, 'end': end }\n",
    "                else:                      # NEGATIVE STRAND\n",
    "                    if  int(dat[3]) - 60 < 60:\n",
    "                        start = 0\n",
    "                    else:\n",
    "                        start = int(dat[3]) - 60\n",
    "                    if start not in geneLookUp:\n",
    "                        geneLookUp[chrom][start] = geneName\n",
    "                    else:\n",
    "                        print('Duplicate start minus strand ', geneName, geneLookUp[chrom][start]) \n",
    "\n",
    "                    if not int(dat[4]) + 72 > S288C_chromSizes[chrom]:\n",
    "                        end = int(dat[4]) + 72\n",
    "                    else:\n",
    "                        end = S288C_chromSizes[chrom]\n",
    "                    # add final gene info for positive strand \n",
    "                    if geneName not in gffDict[chrom]:\n",
    "                        gffDict[chrom][geneName] = {'strand' : '-', 'start': start, 'end': end }\n",
    "\n",
    "\n",
    "# write bed file for use with Samtools  -l flag, start and end positions have been adjust -72 from start and + 60 to end\n",
    "# track name=yeast_genes description=\"All S.cerevisiae Gene locations\"\n",
    "with open('S288C-Genes.bed', 'w') as out:\n",
    "    out.write('track name=yeast_genes description=\"All S.cerevisiae Gene locations\"\\n')\n",
    "    for chrom in gffDict.keys():\n",
    "        for gene, dat in gffDict[chrom].items():\n",
    "            out.write(f'{chrom} {gffDict[chrom][gene][\"start\"]} {gffDict[chrom][gene][\"end\"]} {gene} {gffDict[chrom][gene][\"strand\"]}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a dictionary for bed file\n",
    "genes = {}\n",
    "with open('S288C-Genes.bed', 'r') as f:\n",
    "    f.readline()\n",
    "    for ln in f:\n",
    "        dat = ln.rstrip().split()\n",
    "        if dat[3] not in genes:\n",
    "            genes[dat[3]] = [ dat[1], dat[2], dat[4]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read sample SAM files and produce a count of read start positions that fall within -72 bases upstream of gene start and +60 of gene end.\n",
    "#### I have created a modified bed file with the start and end positions reflecting the -72 and + 60 windows.\n",
    "\n",
    "For example\n",
    "\n",
    "New (plus strand):   ref|NC_001133| 263 709 YAL069W +       ,     Old (from GFF): ref|NC_001133|\t335\t649\t\t+\n",
    "\n",
    "New (minus strand):  ref|NC_001133| 1747 2241 YAL068C -     ,     Old (from GFF): ref|NC_001133|\t1807\t2169\t-"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# YPS chromosome sizes, Need these values for the interval tree\n",
    "YPS1009_chromSizes = {}\n",
    "with open('/mnt/bigdata/linuxhome/mplace/data/reference/YPS1009/YPS1009.chromsizes', 'r') as f:\n",
    "    for ln in f:\n",
    "        chrom, length = ln.rstrip().split()\n",
    "        if chrom not in YPS1009_chromSizes:\n",
    "            YPS1009_chromSizes[chrom] = int(length)\n",
    "\n",
    "print(YPS1009_chromSizes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create information dictionaries for YPS1009\n",
    "gff = '/mnt/bigdata/linuxhome/mplace/data/reference/YPS1009/YPS1009.gff'\n",
    "geneLookUp = {}    # dictionary of dictionaries key = chrom : {key = the -72 position of gene value = gene name }\n",
    "gffDict = {}       # dictionary of dictionaries key = chrom : {key = {start : 0, end : 0, strand : '+'}\n",
    "\n",
    "with open(gff, 'r') as g:\n",
    "    for line in g:\n",
    "        if line.startswith('#'):   # skip comment rows\n",
    "            continue\n",
    "        else:\n",
    "            dat = line.split('\\t')\n",
    "            chrom = dat[0]\n",
    "            if chrom not in YPS1009_chromSizes:\n",
    "                continue\n",
    "            if chrom not in geneLookUp:\n",
    "                geneLookUp[chrom] = {}\n",
    "            # add chrom to gffDict\n",
    "            if chrom not in gffDict:\n",
    "                gffDict[chrom] = {}\n",
    "            # identify genes\n",
    "            if dat[2] == 'gene':\n",
    "                geneName = re.sub('ID=', '', dat[8].split(';')[0])\n",
    "                # identify strand\n",
    "                if dat[6] == '+':          # POSIIVE STRAND\n",
    "                    if int(dat[3]) < 72:          \n",
    "                        start = 0\n",
    "                    else:\n",
    "                        start = int(dat[3]) - 72\n",
    "                    if start not in geneLookUp:\n",
    "                        geneLookUp[chrom][start] = geneName\n",
    "                    else:\n",
    "                        print('Duplicate start ', geneName, geneLookUp[chrom][start])\n",
    "\n",
    "                    if not int(dat[4]) + 60 > YPS1009_chromSizes[chrom]:\n",
    "                        end   = int(dat[4]) + 60\n",
    "                    else:\n",
    "                        end = YPS1009_chromSizes[chrom]\n",
    "\n",
    "                    # add final gene info for positive strand \n",
    "                    if geneName not in gffDict[chrom]:\n",
    "                        gffDict[chrom][geneName] = {'strand' : '+', 'start': start, 'end': end }\n",
    "                else:                      # NEGATIVE STRAND\n",
    "                    if  int(dat[3]) - 60 < 60:\n",
    "                        start = 0\n",
    "                    else:\n",
    "                        start = int(dat[3]) - 60\n",
    "                    if start not in geneLookUp:\n",
    "                        geneLookUp[chrom][start] = geneName\n",
    "                    else:\n",
    "                        print('Duplicate start minus strand ', geneName, geneLookUp[chrom][start]) \n",
    "\n",
    "                    if not int(dat[4]) + 72 > YPS1009_chromSizes[chrom]:\n",
    "                        end = int(dat[4]) + 72\n",
    "                    else:\n",
    "                        end = YPS1009_chromSizes[chrom]\n",
    "                    # add final gene info for positive strand \n",
    "                    if geneName not in gffDict[chrom]:\n",
    "                        gffDict[chrom][geneName] = {'strand' : '-', 'start': start, 'end': end }\n",
    "\n",
    "\n",
    "# write bed file for use with Samtools  -l flag, start and end positions have been adjust -72 from start and + 60 to end\n",
    "# track name=yeast_genes description=\"All S.cerevisiae Gene locations\"\n",
    "with open('YPS1009-Genes.bed', 'w') as out:\n",
    "    out.write('track name=yeast_genes description=\"All YPS1009 S.cerevisiae Gene locations\"\\n')\n",
    "    for chrom in gffDict.keys():\n",
    "        for gene, dat in gffDict[chrom].items():\n",
    "            out.write(f'{chrom} {gffDict[chrom][gene][\"start\"]} {gffDict[chrom][gene][\"end\"]} {gene} {gffDict[chrom][gene][\"strand\"]}\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Setup interval tree use with S288C.bed file for search the start positions of reads\n",
    "\n",
    "Interval Tree implementation :  https://github.com/moonso/interval_tree\n",
    "\n",
    "from interval_tree import IntervalTree\n",
    "features = [\n",
    "\n",
    "        [20,400,'id01'], \n",
    "\n",
    "        [30,400,'id02'], \n",
    "\n",
    "        [500,700,'id03'], \n",
    "\n",
    "        [1020,2400,'id04'], \n",
    "\n",
    "        [35891,29949,'id05'], \n",
    "\n",
    "        [899999,900000,'id06'], \n",
    "        \n",
    "        [999000,999000,'id07']\n",
    "    ]\n",
    "my_tree = IntervalTree(features, 1, 1000000)\n",
    "\n",
    "print('Ranges between 0 and 10: %s' % my_tree.find_range([0, 10]))\n",
    ">Ranges between 0 and 10: []\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from interval_tree import IntervalTree\n",
    "\n",
    "prevChrom = 'ref|NC_001133|'\n",
    "chromBeds = {}\n",
    "with open('S288C-Genes.bed', 'r') as f:\n",
    "    f.readline()                        # skip header\n",
    "    features = []\n",
    "\n",
    "    for ln in f:\n",
    "        dat = ln.split()                #  ref|NC_001133| 263 709 YAL069W +\n",
    "\n",
    "        if dat[0] == prevChrom:\n",
    "            row = [int(dat[1]), int(dat[2]), dat[3]]\n",
    "            features.append(row)\n",
    "        elif dat[0] != prevChrom:\n",
    "            chromBeds[prevChrom] = IntervalTree(features,1, int(S288C_chromSizes[prevChrom])) # write most recent\n",
    "            if dat[0] not in chromBeds:\n",
    "                chromBeds[dat[0]] = None\n",
    "            features = []\n",
    "            row = [int(dat[1]), int(dat[2]), dat[3]]\n",
    "            features.append(row)\n",
    "            prevChrom = dat[0]\n",
    "    \n",
    "    chromBeds[prevChrom] = IntervalTree(features,1, int(S288C_chromSizes[prevChrom])) # write most recent\n",
    "\n",
    "\n",
    "\n",
    "#chrI = IntervalTree(features, 1, 230218)\n",
    "# ref|NC_001147| 1089955 1090569 YOR396C-A -\n",
    "#print(chromBeds['ref|NC_001133|'].find_range([120153, 120153]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the S288C modified bed file\n",
    "S288CBED = {}\n",
    "with open('S288C-Genes.bed','r') as f:\n",
    "    f.readline()\n",
    "    for ln in f:\n",
    "        dat = ln.rstrip().split()\n",
    "        if dat[0] not in S288CBED:\n",
    "            S288CBED[dat[0]] = {}\n",
    "            if dat[3] not in S288CBED[dat[0]]:\n",
    "                S288CBED[dat[0]][dat[3]] = {'start': int(dat[1]), 'end': int(dat[2]), 'strand':dat[4]}\n",
    "        else:\n",
    "            if dat[3] not in S288CBED[dat[0]]:\n",
    "                S288CBED[dat[0]][dat[3]] = {'start': int(dat[1]), 'end': int(dat[2]), 'strand':dat[4]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "S288CBED"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Count read start positions across All yeast genes\n",
    "\n",
    "SAM field information:\n",
    "\n",
    " QNAME String [!-?A-~]{1,254} Query template NAME\n",
    "\n",
    " FLAG Int [0, 216 − 1] bitwise FLAG\n",
    "\n",
    " RNAME String \\*|[:rname:∧*=][:rname:]* Reference sequence NAME11\n",
    "\n",
    " POS Int [0, 231 − 1] 1-based leftmost mapping POSition\n",
    "\n",
    " MAPQ Int [0, 28 − 1] MAPping Quality\n",
    "\n",
    " CIGAR String \\*|([0-9]+[MIDNSHPX=])+ CIGAR string\n",
    "\n",
    " RNEXT String \\*|=|[:rname:∧*=][:rname:]* Reference name of the mate/next read\n",
    "\n",
    " PNEXT Int [0, 231 − 1] Position of the mate/next read\n",
    "\n",
    " TLEN Int [−231 + 1, 231 − 1] observed Template LENgth\n",
    "\n",
    " SEQ String \\*|[A-Za-z=.]+ segment SEQuence\n",
    "\n",
    " QUAL String [!-~]+ ASCII of Phred-scaled base QUALity+33"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MINUS STRAND"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# **************************************   MINUS STRAND\n",
    "\n",
    "#for aln in glob.glob('alignments/S288C/*-sorted.sam'):\n",
    "    \n",
    "#    outName = re.sub('-sorted.sam', '-RiboSeq-counts.txt', aln)\n",
    "#    print('processing ', aln, outName)\n",
    "    \n",
    "# set up results dictionary for the counting of start positions\n",
    "results = {}\n",
    "for chromName in chromBeds.keys():\n",
    "    if chromName not in results:\n",
    "        results[chromName] = {}   \n",
    "\n",
    "# open sam file , MAKE SURE THE SAM FILE IS SORTED (samtools sort)\n",
    "# we are interested in columns  2, 3, 4 (FLAG, RNAME, POS)\n",
    "# we use the NM tag to decide if there is an exact match\n",
    "# NM Edit distance to the reference, so NM:i:0  is an exact match anything else is not.\n",
    "#with open(aln, 'r') as f:\n",
    "with open('alignments/S288C/ANEU-LOG-30-sorted.sam', 'r') as f:\n",
    "    for line in f:\n",
    "        # skip header information\n",
    "        if not line.startswith('@'):\n",
    "            read = line.split('\\t')\n",
    "            if read[1] == '4':             # these reads are unmapped\n",
    "                continue\n",
    "            elif read[1] == '16':           # mapped in forward direction\n",
    "                chrom    = read[2]\n",
    "                readStart = int(read[3])\n",
    "                tags     = read[11:]       # section contains NM tag\n",
    "                for tag in tags:           # find NM tag\n",
    "                    if tag.startswith('NM:'):\n",
    "                        myTag = tag.split('NM:i:')[1]   \n",
    "                        if myTag == '0':\n",
    "                            gene = chromBeds[chrom].find_range([readStart, readStart])   # which gene does this read start in\n",
    "                            if gene is not None:                                       \n",
    "                                for g in gene:                                         # this is a list, make sure we have the proper strand \n",
    "                                    if S288CBED[chrom][g]['strand'] == '-':\n",
    "                                        #out.write(f'{chrom},  {readStart}, {tag}, {myTag}, {g}\\n')\n",
    "                                        #print( 'start ' , S288CBED[chrom][g]['start'], 'end', S288CBED[chrom][g]['end'],\n",
    "                                        #                    'strand', S288CBED[chrom][g]['strand'], 'pos', readStart, 'cnts', 1 )\n",
    "                                        if g not in results[chrom]:\n",
    "                                            results[chrom][g] = {'start' : S288CBED[chrom][g]['start'], 'end': S288CBED[chrom][g]['end'],\n",
    "                                                            'strand': S288CBED[chrom][g]['strand'], 'pos' : {readStart : 1 }}\n",
    "                                        else:\n",
    "                                            if readStart in results[chrom][g]['pos']:\n",
    "                                                results[chrom][g]['pos'][readStart] += 1 \n",
    "                                            else:\n",
    "                                                results[chrom][g]['pos'][readStart] = 1     \n",
    "                                                                    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results['ref|NC_001134|']['YBL039C']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Write table of start position counts\n",
    "\n",
    "#with open(outName, 'w') as out:\n",
    "with open('MINUS-TEST.txt', 'w') as out:\n",
    "    for c in results.keys():\n",
    "        for gene in results[c].keys():\n",
    "            sortedPos = list(results[c][gene]['pos'].keys())\n",
    "            startPos = results[c][gene]['start']\n",
    "            endPos   = results[c][gene]['end']\n",
    "            geneLength = endPos = startPos\n",
    "            strand   = results[c][gene]['strand']\n",
    "            sortedPos.sort()\n",
    "            #print(c, gene, results[c][gene]['start'], results[c][gene]['end'], results[c][gene]['strand'] )\n",
    "            outCodon = []         \n",
    "            outPos = []           # genome position\n",
    "            outCnt = []\n",
    "              \n",
    "            for p in range(results[c][gene]['start'], results[c][gene]['end']):\n",
    "                outPos.append(str(p))\n",
    "                if p in results[c][gene]['pos']:\n",
    "                    outCnt.append(str(results[c][gene]['pos'][p]))\n",
    "                else:\n",
    "                    outCnt.append('0')\n",
    "            \n",
    "            # create the codon positions starting at -71\n",
    "            for i, tmp in enumerate(outPos, start=-71):\n",
    "                outCodon.append(str(i))\n",
    "            \n",
    "            out.write(f'{c}\\n')\n",
    "            outPos.reverse()\n",
    "            posLine   = f'Genome_Position ' + ' '.join(outPos) + '\\n'\n",
    "            out.write(posLine)\n",
    "            \n",
    "            codonLine = f'{results[c][gene][\"strand\"]} ' + ' '.join(outCodon) + '\\n'\n",
    "            out.write(codonLine)\n",
    "            \n",
    "            outCnt.reverse()\n",
    "            countLine = f'{gene} ' + ' '.join(outCnt)   + '\\n'\n",
    "            out.write(countLine)\n",
    "            \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PLUS STRAND NEXT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PLUS STRAND\n",
    "\n",
    "for aln in glob.glob('alignments/S288C/*-sorted.sam'):\n",
    "    \n",
    "    outName = re.sub('-sorted.sam', '-RiboSeq-counts.txt', aln)\n",
    "    print('processing ', aln, outName)\n",
    "    \n",
    "    # set up results dictionary for the counting of start positions\n",
    "    results = {}\n",
    "    for chromName in chromBeds.keys():\n",
    "        if chromName not in results:\n",
    "            results[chromName] = {}   \n",
    "\n",
    "    # open sam file , MAKE SURE THE SAM FILE IS SORTED (samtools sort)\n",
    "    # we are interested in columns  2, 3, 4 (FLAG, RNAME, POS)\n",
    "    # we use the NM tag to decide if there is an exact match\n",
    "    # NM Edit distance to the reference, so NM:i:0  is an exact match anything else is not.\n",
    "    with open(aln, 'r') as f:\n",
    "        for line in f:\n",
    "            # skip header information\n",
    "            if not line.startswith('@'):\n",
    "                read = line.split('\\t')\n",
    "                if read[1] == '4':             # these reads are unmapped\n",
    "                    continue\n",
    "                elif read[1] == '0':           # mapped in forward direction\n",
    "                    chrom    = read[2]\n",
    "                    readStart = int(read[3])\n",
    "                    tags     = read[11:]       # section contains NM tag\n",
    "                    for tag in tags:           # find NM tag\n",
    "                        if tag.startswith('NM:'):\n",
    "                            myTag = tag.split('NM:i:')[1]   \n",
    "                            if myTag == '0':\n",
    "                                gene = chromBeds[chrom].find_range([readStart, readStart])   # which gene does this read start in\n",
    "                                if gene is not None:                                       \n",
    "                                    for g in gene:                                         # this is a list, make sure we have the proper strand \n",
    "                                        if S288CBED[chrom][g]['strand'] == '+':\n",
    "                                            #out.write(f'{chrom},  {readStart}, {tag}, {myTag}, {g}\\n')\n",
    "                                            #print( 'start ' , S288CBED[chrom][g]['start'], 'end', S288CBED[chrom][g]['end'],\n",
    "                                            #                    'strand', S288CBED[chrom][g]['strand'], 'pos', readStart, 'cnts', 1 )\n",
    "                                            if g not in results[chrom]:\n",
    "                                                results[chrom][g] = {'start' : S288CBED[chrom][g]['start'], 'end': S288CBED[chrom][g]['end'],\n",
    "                                                                'strand': S288CBED[chrom][g]['strand'], 'pos' : {readStart : 1 }}\n",
    "                                            else:\n",
    "                                                if readStart in results[chrom][g]['pos']:\n",
    "                                                    results[chrom][g]['pos'][readStart] += 1 \n",
    "                                                else:\n",
    "                                                    results[chrom][g]['pos'][readStart] = 1     \n",
    "                                                                        \n",
    "\n",
    "### Write table of start position counts\n",
    "\n",
    "    with open(outName, 'w') as out:\n",
    "        for c in results.keys():\n",
    "            for gene in results[c].keys():\n",
    "                sortedPos = list(results[c][gene]['pos'].keys())\n",
    "                sortedPos.sort()\n",
    "                #print(c, gene, results[c][gene]['start'], results[c][gene]['end'], results[c][gene]['strand'] )\n",
    "                outCodon = []\n",
    "                outPos = []\n",
    "                outCnt = []\n",
    "                \n",
    "                codonPos = -72  \n",
    "                for p in range(results[c][gene]['start'], results[c][gene]['end']):\n",
    "                    outCodon.append(str(codonPos))\n",
    "                    codonPos +=1\n",
    "                    outPos.append(str(p))\n",
    "                    if p in results[c][gene]['pos']:\n",
    "                        outCnt.append(str(results[c][gene]['pos'][p]))\n",
    "                    else:\n",
    "                        outCnt.append('0')\n",
    "                    \n",
    "                out.write(f'{c}\\n')\n",
    "                posLine   = f'Genome_Position ' + ' '.join(outPos) + '\\n'\n",
    "                out.write(posLine)\n",
    "                codonLine = f'{results[c][gene][\"strand\"]} ' + ' '.join(outCodon) + '\\n'\n",
    "                out.write(codonLine)\n",
    "                countLine = f'{gene} ' + ' '.join(outCnt)   + '\\n'\n",
    "                out.write(countLine)\n",
    "                \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Samtools mpileup -- Previously run, but not really what was required."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# samtools mpileup\n",
    "\n",
    "#samtools mpileup -l regions.bed -f ~/data/reference/S288C_reference_genome_R64-1-1_20110203/S288C_reference_sequence_R64-1-1_20110203.fsa /\n",
    "# ANEU-LOG-30_S2_R1_001-sorted.sam > ANEU-LOG-30_pileup.txt\n",
    "# I created a bed file for all gene positions in YPS1009 and S288C and used this with pileup, this will only output the regions we are interested in viewing.\n",
    "\n",
    "\n",
    "# create pileup output directories\n",
    "pileupPath = parentDir + 'pileup/'\n",
    "os.mkdir(pileupPath)\n",
    "os.mkdir(pileupPath + 'YPS1009')\n",
    "os.mkdir(pileupPath + 'S288C')\n",
    "\n",
    "\n",
    "# Create samtools pileup input file for YPS1009\n",
    "yps1009Ref = '/home/glbrc.org/mplace/data/reference/YPS1009/YPS1009.fasta'\n",
    "tmpPath = pileupPath + 'YPS1009/'\n",
    "with open('pileUp_YPS1009_input.txt', 'w') as out:\n",
    "    for sam in glob.glob(parentDir + 'alignments/YPS1009/*.bam'):\n",
    "        sampleName = re.sub('.bam', '-pileup.txt', os.path.basename(sam))\n",
    "        out.write(sam + ' ' + tmpPath + sampleName + ' ' + yps1009Ref + '\\n')\n",
    "out.close\n",
    "\n",
    "\n",
    "# Create samtools pileup input file for YPS1009\n",
    "yps1009Ref = '/home/glbrc.org/mplace/data/reference/YPS1009/YPS1009.fasta'\n",
    "tmpPath = pileupPath + 'YPS1009/'\n",
    "with open('pileUp_YPS1009_input.txt', 'w') as out:\n",
    "    for sam in glob.glob(parentDir + 'alignments/YPS1009/*.bam'):\n",
    "        sampleName = re.sub('.bam', '-pileup.txt', os.path.basename(sam))\n",
    "        out.write(sam + ' ' + tmpPath + sampleName + ' ' + yps1009Ref + '\\n')\n",
    "out.close\n",
    "\n",
    "\n",
    "# Create pileup.sh input file for S288C\n",
    "s288cRef = '/mnt/bigdata/linuxhome/mplace/data/reference/S288C_reference_genome_R64-1-1_20110203/S288C_reference_sequence_R64-1-1_20110203.fsa'\n",
    "tmpPath = pileupPath + 'S288C/'\n",
    "with open('pileUp_S288C_input.txt', 'w') as out:\n",
    "    for bam in glob.glob(parentDir + 'alignments/S288C/*.bam'):\n",
    "        sampleName = re.sub('.bam', '-pileup.txt', os.path.basename(bam))\n",
    "        out.write(bam + ' ' + tmpPath + sampleName + ' ' + s288cRef + '\\n')\n",
    "out.close\n",
    "# Setup pileup.sh job for S288C\n",
    "with open('s288cPileUp.submit', 'w') as submit:\n",
    "    submit.write( \"Universe                 = vanilla\\n\" )\n",
    "    submit.write( \"Executable               = runS288CPileUp.sh\\n\")\n",
    "    submit.write( \"Arguments                = $(bam) $(pileup) $(ref)\\n\")\n",
    "    submit.write( \"Error                    = s288c_pileup.submit.err\\n\")\n",
    "    submit.write( \"Log                      = s288c_pileup.submit.log\\n\")  \n",
    "    submit.write( \"Requirements             = OpSysandVer == \\\"CentOS7\\\"\\n\")\n",
    "    submit.write( \"Queue bam, pileup, ref from pileUp_S288C_input.txt\\n\" )\n",
    "submit.close()\n",
    "\n",
    "# write shell script to run pileup\n",
    "with open('runS288CPileUp.sh', 'w') as out:\n",
    "    out.write(\"#!/bin/bash\\n\")\n",
    "    out.write(\"source /opt/bifxapps/miniconda3/etc/profile.d/conda.sh\\n\")\n",
    "    out.write(\"unset PYTHONPATH\\n\")  \n",
    "    out.write(\"conda activate /home/glbrc.org/mplace/.conda/envs/riboSeq\\n\")\n",
    "    out.write(f\"samtools mpileup -f $3 $1 > $2\\n\")\n",
    "    out.write(\"conda deactivate\")\n",
    "out.close()\n",
    "\n",
    "os.chmod('runS288CPileUp.sh', 0o0777)\n",
    "\n",
    "#YPS1009\n",
    "with open('yps1009PileUp.submit', 'w') as submit:\n",
    "    submit.write( \"Universe                 = vanilla\\n\" )\n",
    "    submit.write( \"Executable               = runPileUp.sh\\n\")\n",
    "    submit.write( \"Arguments                = $(bam) $(pileup) $(ref) $\\n\")\n",
    "    submit.write( \"Error                    = yps1009_pileup.submit.err\\n\")\n",
    "    submit.write( \"Log                      = yps1009_pileup.submit.log\\n\")  \n",
    "    submit.write( \"Requirements             = OpSysandVer == \\\"CentOS7\\\"\\n\")\n",
    "    submit.write( \"Queue bam, pileup, ref from pileUp_YPS1009_input.txt\\n\" )\n",
    "submit.close()\n",
    "\n",
    "# write shell script to run cutadapt\n",
    "with open('runPileUp.sh', 'w') as out:\n",
    "    out.write(\"#!/bin/bash\\n\")\n",
    "    out.write(\"source /opt/bifxapps/miniconda3/etc/profile.d/conda.sh\\n\")\n",
    "    out.write(\"unset PYTHONPATH\\n\")  \n",
    "    out.write(\"conda activate /home/glbrc.org/mplace/.conda/envs/riboSeq\\n\")\n",
    "    out.write(f\"samtools mpileup -f $3 $1 > $2\\n\")\n",
    "    out.write(\"conda deactivate\")\n",
    "out.close()\n",
    "\n",
    "os.chmod('runPileUp.sh', 0o0777)\n",
    "\n",
    "# open pileup file and create table\n",
    "sample = {}   # key = chrom : geneName { positions: [],  Counts : []}\n",
    "\n",
    "with open('pileup/S288C/ANEU-LOG-30_S2_R1_001-pileup.txt', 'r') as f:\n",
    "    for ln in f:\n",
    "        dat = ln.split('\\t')\n",
    "        if dat[0] not in sample:            # we encounter a new chromosome\n",
    "            gene = chromBeds[dat[0]].find_range([int(dat[1]), int(dat[1])])\n",
    "            sample[dat[0]] = {}\n",
    "            for gn in gene:                  # new genes added\n",
    "                sample[dat[0]][gn] = {'start' : genes[gn][0], 'end': genes[gn][1], 'strand':genes[gn][2], 'pos' : [dat[1]], 'cnts':[dat[3]] }\n",
    "            print(dat[0], gene,dat[1])\n",
    "        else:\n",
    "            gene = chromBeds[dat[0]].find_range([int(dat[1]), int(dat[1])])      # find gene\n",
    "\n",
    "            for gn in gene:\n",
    "                if gn not in sample[dat[0]]:\n",
    "                    sample[dat[0]][gn] = {'start' : genes[gn][0], 'end': genes[gn][1], 'strand':genes[gn][2], 'pos' : [dat[1]], 'cnts':[dat[3]] }\n",
    "                else:\n",
    "                    sample[dat[0]][gn]['pos'].append(dat[1])\n",
    "                    sample[dat[0]][gn]['cnts'].append(dat[3])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "riboSeq",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
